File: app/src/main/java/com/keylesspalace/tusky/pager/ImagePagerAdapter.java
package com.keylesspalace.tusky.pager;

import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;

import com.keylesspalace.tusky.entity.Attachment;
import com.keylesspalace.tusky.fragment.ViewMediaFragment;

import java.util.List;
import java.util.Locale;

public final class ImagePagerAdapter extends FragmentPagerAdapter {

    private List<Attachment> attachments;
    private int initialPosition;

    public ImagePagerAdapter(FragmentManager fragmentManager, List<Attachment> attachments, int initialPosition) {
        super(fragmentManager);
        this.attachments = attachments;
        this.initialPosition = initialPosition;
    }

    @Override
    public Fragment getItem(int position) {
        if (position >= 0 && position < attachments.size()) {
            return ViewMediaFragment.newInstance(attachments.get(position), position == initialPosition);
        } else {
            return null;
        }
    }

    @Override
    public int getCount() {
        return attachments.size();
    }

    @Override
    public CharSequence getPageTitle(int position) {
        return String.format(Locale.getDefault(), "%d/%d", position + 1, attachments.size());
    }
}

File: app/src/main/java/com/keylesspalace/tusky/pager/ImagePagerAdapter.kt
package com.keylesspalace.tusky.pager

import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import com.keylesspalace.tusky.ViewMediaAdapter
import com.keylesspalace.tusky.entity.Attachment
import com.keylesspalace.tusky.fragment.ViewMediaFragment
import java.lang.ref.WeakReference

class ImagePagerAdapter(
        activity: FragmentActivity,
        private val attachments: List<Attachment>,
        private val initialPosition: Int
) : ViewMediaAdapter(activity) {

    private var didTransition = false
    private val fragments = MutableList<WeakReference<ViewMediaFragment>?>(attachments.size) { null }

    override fun getItemCount() = attachments.size

    override fun createFragment(position: Int): Fragment {
        if (position >= 0 && position < attachments.size) {
            // Fragment should not wait for or start transition if it already happened but we
            // instantiate the same fragment again, e.g. open the first photo, scroll to the
            // forth photo and then back to the first. The first fragment will try to start the
            // transition and wait until it's over and it will never take place.
            val fragment = ViewMediaFragment.newInstance(
                    attachment = attachments[position],
                    shouldStartPostponedTransition = !didTransition && position == initialPosition
            )
            fragments[position] = WeakReference(fragment)
            return fragment
        } else {
            throw IllegalStateException()
        }
    }

   override fun onTransitionEnd(position: Int) {
        this.didTransition = true
        fragments[position]?.get()?.onTransitionEnd()
    }
}
